src/gui/qgsmetadatawidget.cpp

```diff
@@ -0,0 +1,806 @@
+  }
+  layerMetadata.setLicenses( licenses );
+
+  // Rights
+  layerMetadata.setRights( mRightsModel->stringList() );
+
+  // Constraints
+  QList<QgsLayerMetadata::Constraint> constraints;
+  for ( int row = 0 ; row < mConstraintsModel->rowCount() ; row++ )
+  {
+    struct QgsLayerMetadata::Constraint constraint = QgsLayerMetadata::Constraint();
+    constraint.type = mConstraintsModel->item( row, 0 )->text();
+    constraint.constraint = mConstraintsModel->item( row, 1 )->text();
+    constraints.append( constraint );
+  }
+  layerMetadata.setConstraints( constraints );
+
+  // CRS
+  if ( selectionCrs->crs().isValid() )
+  {
+    layerMetadata.setCrs( selectionCrs->crs() );
+  }
+
+  // Links
+  QList<QgsLayerMetadata::Link> links;
+  for ( int row = 0 ; row < mLinksModel->rowCount() ; row++ )
+  {
+    struct QgsLayerMetadata::Link link = QgsLayerMetadata::Link();
+    link.name = mLinksModel->item( row, 0 )->text();
+    link.type = mLinksModel->item( row, 1 )->text();
+    link.url = mLinksModel->item( row, 2 )->text();
+    link.description = mLinksModel->item( row, 3 )->text();
+    link.format = mLinksModel->item( row, 4 )->text();
+    link.mimeType = mLinksModel->item( row, 5 )->text();
+    link.size = mLinksModel->item( row, 6 )->text();
+    links.append( link );
+  }
+  layerMetadata.setLinks( links );
+
+  // History
+  layerMetadata.setHistory( mHistoryModel->stringList() );
+}
+
+bool QgsMetadataWidget::checkMetadata() const
+{
+  QgsLayerMetadata metadata = QgsLayerMetadata();
+  saveMetadata( metadata );
+  QgsNativeMetadataValidator validator;
+  QList<QgsMetadataValidator::ValidationResult> validationResults;
+  bool results = validator.validate( metadata, validationResults );
+
+  QString errors;
+  if ( results == false )
+  {
+    errors = QStringLiteral();
+    for ( const QgsMetadataValidator::ValidationResult &result : validationResults )
+    {
+      errors += QLatin1String( "<b>" ) % result.section;
+      if ( ! result.identifier.isNull() )
+      {
+        errors += QLatin1String( " " ) % QVariant( result.identifier.toInt() + 1 ).toString();
+      }
+      errors += QLatin1String( "</b>: " ) % result.note % QLatin1String( "<br />" );
+    }
+  }
+  else
+  {
+    errors = QStringLiteral( "Ok, it seems valid." );
+  }
+
+  QString myStyle = QgsApplication::reportStyleSheet();
+  myStyle.append( QStringLiteral( "body { margin: 10px; }\n " ) );
+  resultsCheckMetadata->clear();
+  resultsCheckMetadata->document()->setDefaultStyleSheet( myStyle );
+  resultsCheckMetadata->setHtml( errors );
+
+  return results;
+}
+
+QMap<QString, QString> QgsMetadataWidget::parseLanguages()
+{
+  QString path = QDir( QgsApplication::metadataPath() ).absoluteFilePath( QString( "country_code_ISO_3166.csv" ) );
+  QFile file( path );
+  if ( !file.open( QIODevice::ReadOnly ) )
+  {
+    QgsDebugMsg( QString( "Error while opening the CSV file: %1, %2 " ).arg( path, file.errorString() ) );
```

Comment:
shouldn't this return here?

Reply:
I'm returning an empty QStringList or QMap. Or Are you talking about the error message?